### PR TITLE
Dev br

### PR DESCRIPTION
**Pull request by LaurentiuTusa**
### Summary
This PR implements the Bank ATM system

###Implementation details
- loop for user interaction with the console
- 3 files for development + 1 file for tests
- 3 classes: Person, BankAccount, Transaction
- each person can have multiple accounts
- each account can have multiple transactions
- daily limit of 5000 for withdrawals: compute the sum of all withdrawals from the current day and compare with the requested amount
- input validations for amount

### CODE REVIEW

File: bank_atm.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    case option
+        when "1" #deposit
+            puts "Insert your name:"
+            input_name = gets.chomp
+            puts "Insert PIN:"
+            input_pin = gets.chomp
+
+            #verify pin
+            current_user = persons.find {|person| person.name == input_name}
```
Comment (VictorMotogna) [inline bank_atm.rb]: I notice that the current_user & the PIN check part is the same in multiple areas. I think you should extract these (including the conditions & checks) into a separate method, that can be reused

File: bank_atm.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    puts "3. Create account"
+    puts "4. View all accounts"
+    puts "5. Quit"
+
+    option = gets.chomp.downcase
+    break if option == "5"
+    puts "You selected #{option}"
+
+    case option
+        when "1" #deposit
```
Comment (VictorMotogna) [inline bank_atm.rb]: I think the indentation here is wrong, it's good practice to do:
```ruby
case option
when 1
  something
when 2
  something else
# ...
```

so the the case & when are aligned

File: models.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
+require 'date'
+
+class Person
```
Comment (bogdan-boer) [inline models.rb]: Nice 👍

File: bank_atm.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    puts "2. Withdraw funds"
+    puts "3. Create account"
+    puts "4. View all accounts"
+    puts "5. Quit"
+
+    option = gets.chomp.downcase
+    break if option == "5"
+    puts "You selected #{option}"
+
+    case option
```
Comment (VictorMotogna) [inline bank_atm.rb:104]: I think a better option here would be to extract the logic into separate methods:
```ruby
case option
when 1
  deposit_funds
when 2
  withdraw_funds
when 3
  create_account
# ... etc
```

This way, the logic can be found in methods, easier to find them and the switch won't get very long

File: bank_atm.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    puts "Select an option: "
+    #Options
+    puts "1. Deposit funds"
+    puts "2. Withdraw funds"
+    puts "3. Create account"
+    puts "4. View all accounts"
+    puts "5. Quit"
+
+    option = gets.chomp.downcase
+    break if option == "5"
```
Comment (VictorMotogna) [inline bank_atm.rb:101]: I think you would probably want a validation to check if the user enters something else besides 1, 2, 3, 4 or 5

File: models.rb
Comment (bogdan-boer) [inline models.rb:1]: Usually you don't have multiple class definitions in the same file, I would create a folder named `models` and for each class I would create a new file

File: bank_atm.rb (hunk 4 above)
Comment (bogdan-boer) [inline bank_atm.rb:104]: This method is kinda hard to follow, for each case create a method. Then it would look smth like this 
```
case option
  when "1":
    deposit_money
  when "2":
     withdraw_money
....
```

File: bank_atm.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            puts "Insert PIN:"
+            input_pin = gets.chomp
+
+            #verify pin
+            current_user = persons.find {|person| person.name == input_name}
+            if current_user == nil 
+                puts "No user found with that name"
+
+            #find the bank account
+            elsif bank_acc = current_user.bank_accounts.find {|account| account.pin_matches?(input_pin) }
```
Comment (bogdan-boer) [inline bank_atm.rb]: I see this is a repeated part of code, you should create a method with that and reuse it

File: models.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      @balance = balance
+      @pin = pin
+      @person = person
+      @transactions = []
+      person.add_bank_account(self)
+    end
+  
+    def pin_matches?(input_pin)
+        @pin == input_pin
+      end
```
Comment (VictorMotogna) [inline models.rb]: You have an indentation issue here

File: models.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    def add_bank_account(account)
+      @bank_accounts << account
+    end
+  
+    def remove_bank_account(account)
+      @bank_accounts.delete(account)
+    end
+
+    def to_string
+        return "name: #{@name}, job: #{@job}, email: #{email}, bank_accounts: #{@bank_accounts.length}"
```
Comment (VictorMotogna) [inline models.rb]: No need to use `return` in Ruby in this case. Methods in Ruby automatically return the last row from the method. For example:
```ruby
def string
  something = "This is a string"

  something
end
```
,
```ruby
def string
  something = "This is a string"

  return something
end
```
,
```ruby
def string
  return "This is a string"
end
```
and
```ruby
def string
  "This is a string"
end
```

will all have the same output

File: models.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                if Validations.validate_withdrawal_daily_limit(amount, @transactions)
+                    @balance -= amount
+                    #add the transaction
+                    Transaction.new(amount, "withdrawal", self)
+                    puts "You withdrew #{amount}. Your new balance is #{@balance}"
+                else
+                    puts "You have exceeded your daily withdrawal limit"
+                end
+            else
+            puts "Insufficient funds"
```
Comment (VictorMotogna) [inline models.rb]: You have an indentation issue here

File: models.rb [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            puts "Invalid amount"
+        end
+    end
+
+    def add_transaction(transaction)
+        @transactions << transaction
+    end
+
+    def to_string
+        return "balance: #{@balance}, pin: #{@pin}, person: #{@person.name}, transactions: #{@transactions.length}"
```
Comment (VictorMotogna) [inline models.rb]: same here, no need to use return

File: spec/validations_spec.rb [hunk 11]
```diff
@@ -0,0 +1,127 @@
+require_relative "../validations"
+require_relative "../models"
+
+#rspec file for testing the validations.rb file
+RSpec.describe Validations do
+
```
Comment (VictorMotogna) [inline spec/validations_spec.rb]: you could remove this newline

File: spec/validations_spec.rb [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            expect(p1.bank_accounts.length).to eq(0)
+            b = BankAccount.new("1234", p1, 7000)
+            expect(p1.bank_accounts.length).to eq(1)
+        end
+    end
+end
+
+#rspec file for testing the BankAccount class
+RSpec.describe BankAccount do
+
```
Comment (VictorMotogna) [inline spec/validations_spec.rb]: you could remove this newline

File: models.rb (hunk 7 above)
Comment (bogdan-boer) [inline models.rb]: 👀

File: spec/validations_spec.rb [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            p1 = Person.new("Alex", "Doctor", "alex@yahoo.com")
+            expect(p1.bank_accounts.length).to eq(0)
+            b = BankAccount.new("1234", p1, 7000)
+            expect(p1.bank_accounts.length).to eq(1)
+        end
+    end
+end
+
+#rspec file for testing the BankAccount class
+RSpec.describe BankAccount do
```
Comment (VictorMotogna) [inline spec/validations_spec.rb]: I think you could extract this into separate files, something like: `spec/validations/bank_account.rb` & `spec/validations/person.rb`

File: models.rb [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    def pin_matches?(input_pin)
+        @pin == input_pin
+      end
+
+    def deposit(amount)
+      if Validations.validate_amount(amount)
+        @balance += amount
+        #add the transaction
+        Transaction.new(amount, "deposit", self)
+        puts "You deposited #{amount}. Your new balance is #{@balance}"
```
Comment (bogdan-boer) [inline models.rb]: You should not print things in logic classes, this should be done in main

File: validations.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
+module Validations
+    def Validations.validate_amount(amount)
```
Comment (bogdan-boer) [inline validations.rb]:
```suggestion
    def self.validate_amount(amount)
```

File: validations.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        if amount > 0
+            return true
+        else
+            return false
+        end
+    end
+
+    def Validations.validate_withdrawal(amount, balance)
```
Comment (bogdan-boer) [inline validations.rb]: Same as above

File: validations.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+        if amount > balance
+            return false
+        else
+            return true
+        end
+    end
+
+    def Validations.validate_withdrawal_daily_limit(amount, transactions)
```
Comment (bogdan-boer) [inline validations.rb]: Same as above

File: validations.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
+module Validations
+    def Validations.validate_amount(amount)
+        if amount > 0
```
Comment (VictorMotogna) [inline validations.rb]: You could do
```ruby
def Validations.validate_amount(amount)
  amount > 0
end
```

Comment (VictorMotogna) [reply validations.rb]: Also, there is a Ruby method called `positive?` that does the same thing

File: validations.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+            return true
+        else
+            return false
+        end
+    end
+
+    def Validations.validate_withdrawal(amount, balance)
+        if amount > balance
```
Comment (VictorMotogna) [inline validations.rb]: Same here and I noticed similar below (line 25). `amount > balance` is already a boolean, true or false, so you can return the value of `amount > balance`. It's obsolete to do:
```ruby
if boolean
  return true
else
  return false
end
```

when you can do `return boolean`

File: validations.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
+module Validations
```
Comment (VictorMotogna) [inline validations.rb:1]: After seeing how you implemented this validations module, I think you could've moved these methods into the BankAccount. It makes more sense to have them as `bank_account_instance.validate_withdrawal`, as withdrawal is an action on a BankAccount instance

You can also keep them separate, but find better naming (like AccountValidations), note it as a class and use instance methods, like:
```ruby
class AccountValidations # or BankValidations
  class << self
    def validate_amount(amount)
      # logic
    end
  end
end

# and call it using
AccountValidations.validate_amount(10)
```

File: validations.rb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+                daily_total += transaction.amount
+            end
+        end
+        if daily_total + amount > 5000
+            return false
+        else
+            return true
+        end
+    end
+end
```
Comment (VictorMotogna) [inline validations.rb:23]: I noticed that these validations always return true or false. This is an ok approach, but you could also do something like:
```ruby
def validate_something
  raise "Custom error message" if invalid
  # or even better
  raise StandardError.new "Custom error message" if invalid
end
```

after that, in main, you could do something like
```ruby
def do_something
  # logic to do something
rescue StandardError => e
  puts e # this will print "Custom error message"
end
```

Comment (VictorMotogna) [reply validations.rb:23]: In specs, you can also do
```ruby
  expect { subject }.to raise_error(StandardError, "Custom error message")
```

File: bank_atm.rb [hunk 22]
```diff
@@ -0,0 +1,116 @@
+$LOAD_PATH << '.'
+require 'validations'
+require_relative 'models/initializer'
+
+#Begin of demo data
```
Comment (florinionce) [inline bank_atm.rb:5]: Usually for comments, we should add a space after `#`

Comment (florinionce) [reply bank_atm.rb:5]: Also, instead of adding comments for the demo data setup, you could extract this into a `setup_demo_data` method, thus you would write code that comments itself :D

File: bank_atm.rb [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+b1 = BankAccount.new("1234", p1, 150)
+b2 = BankAccount.new("5555", p1, 5500)# with 1 widthrawal already
+b3 = BankAccount.new("9999", p2, 7000)
+
+b2.withdraw(4000)
+#End of demo data
+
+#Auxiliary methods:
+def validate_person_name(name, persons)
+    person = persons.find { |person| person.name == name }
```
Comment (florinionce) [inline bank_atm.rb:22]: Please set the indentation to two spaces (/tab) for the next assignment :D

File: bank_atm.rb [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+end
+
+def validate_person_bank_account(bank_accounts, pin)
+    selected_bank_acc = bank_accounts.find { |account| account.pin_matches?(pin) }
+    raise StandardError.new "Invalid PIN" if selected_bank_acc.nil?
+    selected_bank_acc
+end
+
+#Options methods:
+def deposit_funds(persons)
```
Comment (florinionce) [inline bank_atm.rb:34]: you could avoid always passing the `persons` as an argument by setting the persons as an instance variable

File: models/bankAccount.rb [hunk 25]
```diff
@@ -0,0 +1,41 @@
+require_relative '../validations' 
+
+AccountValidations = Validations::AccountValidations
```
Comment (florinionce) [inline models/bankAccount.rb:3]: Why was that? :D

File: models/bankAccount.rb [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+      AccountValidations.validate_withdrawal_daily_limit(amount, @transactions) #check daily limit
+      @balance -= amount 
+      Transaction.new(amount, "withdrawal", self)
+    end
+
+    def add_transaction(transaction)
+      @transactions << transaction
+    end
+
+    def to_string
```
Comment (florinionce) [inline models/bankAccount.rb:38]: usually, in ruby this method is called `to_s`

File: spec/validations_spec.rb [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+                expect { AccountValidations.validate_withdrawal_daily_limit(100, transactions) }.not_to raise_error
+            end
+
+            it "raises an error message if amount is greater than 5000 for the current day" do
+                p_test = Person.new("Alex", "Doctor", "alex@yahoo.com")
+                b_test = BankAccount.new("1234", p_test, 7000)
+                transactions = b_test.transactions
+
+                Transaction.new(4000, "withdrawal", b_test)
+                expect { AccountValidations.validate_withdrawal_daily_limit(1001, transactions) }.to raise_error(StandardError, "You have exceeded your daily withdrawal limit")#5000 is allowed for the day, so we expect false for 5001
```
Comment (florinionce) [inline spec/validations_spec.rb:51]: this sort of comments are not actually needed in my opinion, since it's quite obvious from the context. In this case, from the test's description

**Review by florinionce — APPROVED** [review]
